### ISSUE TITLE
Avoiding `document.write`

### ISSUE THREAD

**Issue by BryceStevenWilley** [krusynth/public-interest-tech-webring#14]
TL;DR: we should consider getting rid of `document.write('<div id="ctwebring"></div>');`

I've been snooping around on my site's console recently and saw this message in Firefox:

> An unbalanced tree was written using document.write() causing data from the network to be reparsed. More information: https://developer.mozilla.org/en-US/docs/Glossary/speculative_parsing

This happens in [webring.js](https://github.com/krusynth/public-interest-tech-webring/blob/923c46cd30535689f6b98fa0c8aba82a8bf82226/webring.js#L3) at the beginning.

Chrome's Lighthouse report also links to [their own docs on the matter](https://developer.chrome.com/docs/lighthouse/best-practices/no-document-write/?utm_source=lighthouse&utm_medium=devtools), and [document.write's docs](https://developer.mozilla.org/en-US/docs/Web/API/Document/write) also list it as deprecated. After reading those pages, I think that the use here is okay? But the fact that Firefox recognizes the write as unbalanced gives me pause, and makes me think we should avoid it.

There are a few options to remedy:

1. Wrap the call to `document.write` in an `if (!document.getElementById("ctwebring"))`, so if you manually add a `<div id="ctwebring"></div>` to your page, `document.write` doesn't get called. The simplest solution, I have a version that works with and without the manual `<div>`, so it shouldn't break anyone.

2. In theory, some combination of `document.createElement()` and `document.insertBefore()` would work. Unfortunately it doesn't seem to work with `document.currentScript`, so we're in the same position the write is trying to solve (not having a known id on the page), and would require some hacky-ness to get around.

**Comment by krusynth** [comment]
I've just gone through all of the other members, including my own site, and yours is the only one throwing this warning (not an error, notably).  So I suspect something else is at play here. Usually the unbalanced tree warning is indicative of a document.write() not writing both the opening and closing tag - but I would also check to see if you have something else not closed properly before you get to that line?

Unfortunately, `document.write` remains stubbornly stable and commonly used pretty much everywhere on the web. It doesn't appear to be actually deprecated (meaning, "browsers: stop supporting this!") just not a _best_ practice.

A workaround could be [something like](https://stackoverflow.com/questions/4537963/what-are-alternatives-to-document-write): 

```
document.write=function(s){
    var scripts = document.getElementsByTagName('script');
    var lastScript = scripts[scripts.length-1];
    lastScript.insertAdjacentHTML("beforebegin", s);
}
```

I'm happy to look at that, but I'm a bit concerned about unpredictable behaviors with assumptions around the script loading order in a dynamic environment.

(And of course, as mentioned, I'm not sure that's what's actually causing your issue.)

**Comment by krusynth** [comment]
Also, forcing everyone to change the code on their sites is not something I'm willing to consider at this point. We need a non-invasive fix, as this code is used on other webrings.

**Comment by BryceStevenWilley** [comment]
> I would also check to see if you have something else not closed properly before you get to that line?

You would be correct; double checking, it looks like I had included my `<script>` tag inside of a `<p>`, something I should have noticed was off. Moving it outside of the `<p>` does remove the warning in Firefox.

> Unfortunately, document.write remains stubbornly stable and commonly used pretty much everywhere on the web. It doesn't appear to be actually deprecated (meaning, "browsers: stop supporting this!") just not a best practice.

That's fair; I'm not 100% up-to-date on all of my front-end stuff, and was having trouble telling if there were any sort of future issues that could pop up with it.

> forcing everyone to change the code on their sites is not something I'm willing to consider at this point. We need a non-invasive fix, as this code is used on other webrings.

100% agree, apologies if my initial post implied anything else. My first suggestion was trying to be backwards compatible, while giving users a bit more flexibility in where the webring showed up.

Closing this now, since it's fixed on my side, and doesn't seem to be an issue elsewhere. Sorry for the unnecessary issue!
